### PR TITLE
curvefs: support memcache cluster

### PR DESCRIPTION
**Pull request by Cyber-SiKu**
1. add define in topology.proto
2. add RegistMemcacheCluster in topology service
3. add define in topology_item
4. add AllocOrGetMemcacheCluster in client/mdsclinet
5. add ListMemcacheCluster in topology service

Signed-off-by: Cyber-SiKu <Cyber-SiKu@outlook.com>

### What problem does this PR solve?

Issue Number: #xxx 

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by Cyber-SiKu** [comment]
recheck

**Comment by Cyber-SiKu** [comment]
recheck

**Comment by Cyber-SiKu** [comment]
recheck

**Comment by Cyber-SiKu** [comment]
recheck

**Comment by Cyber-SiKu** [comment]
recheck

**Comment by Cyber-SiKu** [comment]
recheck

**Comment by ilixiaocui** [comment]
recheck

**Comment by Cyber-SiKu** [comment]
recheck

File: curvefs/proto/topology.proto [hunk 1]
```diff
@@ -458,6 +459,33 @@ message ListTopologyResponse {
...
+
+message RegistMemcacheClusterResponse {
+    required TopoStatusCode statusCode = 1;
+    optional uint32 clusterId = 2;
+}
+
+message ListMemcacheClusterRequest {
+}
+
+message MemcacheClusterInfo {
```
Comment (cw123) [inline curvefs/proto/topology.proto:479]: the message is not same with #2096

File: curvefs/proto/topology.proto [hunk 2]
```diff
@@ -458,6 +459,33 @@ message ListTopologyResponse {
     required ListMetaServerResponse metaservers = 5;
 }
 
+message MemcacheServerInfo {
```
Comment (cw123) [inline curvefs/proto/topology.proto:462]: the message is not same with https://github.com/opencurve/curve/pull/2096

File: curvefs/src/client/rpcclient/mds_client.cpp [hunk 3]
```diff
@@ -413,6 +413,40 @@ bool MdsClientImpl::ListPartition(uint32_t fsID,
     return 0 == rpcexcutor_.DoRPCTask(task, mdsOpt_.mdsMaxRetryMS);
 }
 
+bool MdsClientImpl::AllocOrGetMemcacheCluster(MemcacheClusterInfo* cluster) {
+    auto task = RPCTask {
```
Comment (cw123) [inline curvefs/src/client/rpcclient/mds_client.cpp:417]: add metric for this

File: curvefs/src/client/rpcclient/mds_client.cpp [hunk 4]
```diff
@@ -413,6 +413,40 @@ bool MdsClientImpl::ListPartition(uint32_t fsID,
...
+            LOG(WARNING) << "ListMemcacheCluster fail, errcode = " << ret
+                         << ", errmsg = " << TopoStatusCode_Name(ret);
+            return ret;
+        }
+
+        ::google::protobuf::RepeatedPtrField<MemcacheClusterInfo> clusters =
+            response.memclusters();
+        if (!clusters.empty()) {
+            int randId =
+                (static_cast<int>(butil::fast_rand()) % clusters.size()) + 1;
```
Comment (cw123) [inline curvefs/src/client/rpcclient/mds_client.cpp:438]: why +1 here

File: curvefs/src/mds/topology/topology_item.h [hunk 5]
```diff
@@ -57,7 +58,7 @@ struct ClusterInformation {
     std::map<uint32_t, uint32_t> partitionIndexs;
 
     ClusterInformation() = default;
-    explicit ClusterInformation(const std::string &clusterId)
+    explicit ClusterInformation(const std::string& clusterId)
```
Comment (cw123) [inline curvefs/src/mds/topology/topology_item.h:61]: 这个文件有大量的格式修改，没有办法review，这种仅仅是格式的修改，而且旧的格式问题不大，最好不要去动他，建议格式恢复回去再review。

This file has a lot of format changes, and there is no way to review it. This is only a format modification, and the old format is not a big problem. It is best not to touch it. It is recommended to restore the format and review it.

File: curvefs/proto/topology.proto (hunk 1 above)
Comment (Cyber-SiKu) [reply curvefs/proto/topology.proto:479]: > the message is not same with #2096

#2096 Merge first, fix when merge.

File: curvefs/src/client/rpcclient/mds_client.cpp (hunk 4 above)
Comment (Cyber-SiKu) [reply curvefs/src/client/rpcclient/mds_client.cpp]: > why +1 here

fix

File: curvefs/src/mds/topology/topology_item.h (hunk 5 above)
Comment (Cyber-SiKu) [reply curvefs/src/mds/topology/topology_item.h]: > 这个文件有大量的格式修改，没有办法review，这种仅仅是格式的修改，而且旧的格式问题不大，最好不要去动他，建议格式恢复回去再review。
> 
> This file has a lot of format changes, and there is no way to review it. This is only a format modification, and the old format is not a big problem. It is best not to touch it. It is recommended to restore the format and review it.

fix

**Comment by Cyber-SiKu** [comment]
recheck

File: curvefs/src/mds/topology/topology_manager.cpp [hunk 6]
```diff
@@ -1318,6 +1320,69 @@ TopoStatusCode TopologyManager::UpdatePartitionStatus(
...
+            response->set_statuscode(errorCode);
+            response->set_clusterid(id);
+        }
+    }
+}
+
+void TopologyManager::ListMemcacheCluster(
+    ListMemcacheClusterResponse* response) {
+    std::list<MemcacheCluster> clusterList = topology_->ListMemcacheClusters();
+    if (clusterList.empty()) {
```
Comment (SeanHai) [inline curvefs/src/mds/topology/topology_manager.cpp:1367]: if (!clusterList.empty())

File: curvefs/src/client/rpcclient/mds_client.cpp [hunk 7]
```diff
@@ -413,6 +413,33 @@ bool MdsClientImpl::ListPartition(uint32_t fsID,
...
+        }
+
+        TopoStatusCode ret = response.statuscode();
+        if (ret != TopoStatusCode::TOPO_OK) {
+            LOG(WARNING) << "AllocOrGetMemcacheCluster fail, errcode = " << ret
+                         << ", errmsg = " << TopoStatusCode_Name(ret);
+            return ret;
+        }
+
+        *cluster = response.cluster();
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/client/rpcclient/mds_client.cpp:435]: > *cluster = std::move(*response.mutable_cluster()) ?

image

File: curvefs/src/mds/topology/topology_manager.cpp [hunk 8]
```diff
@@ -1318,6 +1320,69 @@ TopoStatusCode TopologyManager::UpdatePartitionStatus(
     return topology_->UpdatePartitionStatus(partitionId, status);
 }
 
+void TopologyManager::RegistMemcacheCluster(
+    const RegistMemcacheClusterRequest* request,
+    RegistMemcacheClusterResponse* response) {
+    response->set_statuscode(TopoStatusCode::TOPO_OK);
+    // register memcacheCluster as server
+    curve::common::NameLockGuard lock(registMemcacheClusterMutex_,
+                                      request->ShortDebugString());
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/mds/topology/topology_manager.cpp:1329]: > I'm not sure the ShortDebugString are the same when the different order of servers added in request.

fix

File: curvefs/src/mds/topology/topology_manager.cpp [hunk 9]
```diff
@@ -1318,6 +1320,69 @@ TopoStatusCode TopologyManager::UpdatePartitionStatus(
...
+
+    // Guarantee the uniqueness of memcacheServer
+    std::list<MemcacheServer> serverRegisted = topology_->ListMemcacheServers();
```
[the remark this answers is not in the document]
Comment (Cyber-SiKu) [reply curvefs/src/mds/topology/topology_manager.cpp:1332]: > Can one machine serve two clusters?

yep

File: curvefs/src/mds/topology/topology_manager.cpp (hunk 6 above)
Comment (Cyber-SiKu) [reply curvefs/src/mds/topology/topology_manager.cpp:1367]: > if (!clusterList.empty())

fix

File: curvefs/src/mds/topology/topology_storge_etcd.h [hunk 10]
```diff
@@ -86,6 +86,17 @@ class TopologyStorageEtcd : public TopologyStorage {
...
 
+    bool LoadMemcacheCluster(
+        std::unordered_map<MemcacheClusterIdType, MemcacheCluster>*
+            memcacheClusterMap,
+        MemcacheClusterIdType* maxMemCacheClusterId) override;
+    bool StorageMemcacheCluster(const MemcacheCluster& data) override;
+
+    bool StorageFs2MemCluster(FsIdType fsId,
+                              MemcacheClusterIdType memClusterId) override;
+    bool LoadFs2MemCluster(std::unordered_map<FsIdType, MemcacheClusterIdType>*
```
Comment (cw123) [inline curvefs/src/mds/topology/topology_storge_etcd.h:97]: MemCluster is MemcacheCluster

memCluster may be mistaken for memoryCluster

File: curvefs/src/mds/topology/topology_manager.cpp [hunk 11]
```diff
@@ -1318,6 +1320,68 @@ TopoStatusCode TopologyManager::UpdatePartitionStatus(
...
+            TopoStatusCode errorCode =
+                topology_->AddMemcacheCluster(std::move(cluster));
+            response->set_statuscode(errorCode);
+            response->set_clusterid(id);
+        }
+    }
+}
+
+void TopologyManager::ListMemcacheCluster(
+    ListMemcacheClusterResponse* response) {
```
Comment (cw123) [inline curvefs/src/mds/topology/topology_manager.cpp:1364]: add readlock registMemcacheClusterMutex_

File: curvefs/src/mds/topology/topology_manager.cpp [hunk 12]
```diff
@@ -1318,6 +1320,68 @@ TopoStatusCode TopologyManager::UpdatePartitionStatus(
...
+        }
+    } else {
+        response->set_statuscode(
+            TopoStatusCode::TOPO_MEMCACHECLUSTER_NOT_FOUND);
+    }
+}
+
+void TopologyManager::AllocOrGetMemcacheCluster(
+    const AllocOrGetMemcacheClusterRequest* request,
+    AllocOrGetMemcacheClusterResponse* response) {
```
Comment (cw123) [inline curvefs/src/mds/topology/topology_manager.cpp:1379]: add readlock registMemcacheClusterMutex_

File: curvefs/src/mds/topology/topology.cpp [hunk 13]
```diff
@@ -1656,6 +1677,75 @@ bool TopologyImpl::RefreshPartitionIndexOfFS(
...
+
+TopoStatusCode TopologyImpl::AddMemcacheCluster(MemcacheCluster&& data) {
+    WriteLockGuard wlockMemcacheCluster(memcacheClusterMutex_);
+    // storage_ to storage
+    TopoStatusCode ret = TopoStatusCode::TOPO_OK;
+    if (!storage_->StorageMemcacheCluster(data)) {
+        ret = TopoStatusCode::TOPO_STORGE_FAIL;
+    } else {
+        memClusterMap_[data.GetId()] = data;
+        memClusterMap_.insert(std::make_pair(data.GetId(), std::move(data)));
```
Comment (cw123) [inline curvefs/src/mds/topology/topology.cpp:1711]: two line is dup

File: curvefs/src/mds/topology/topology.cpp [hunk 14]
```diff
@@ -1656,6 +1677,75 @@ bool TopologyImpl::RefreshPartitionIndexOfFS(
...
+    ReadLockGuard rlockMemcacheCluster(memcacheClusterMutex_);
+    if (fs2MemCluster_.find(fsId) != fs2MemCluster_.end()) {
+        *cluster = memClusterMap_[fs2MemCluster_[fsId]];
+    } else if (memClusterMap_.empty()) {
+        ret = TopoStatusCode::TOPO_MEMCACHECLUSTER_NOT_FOUND;
+    } else {
+        int randId =
+            static_cast<int>(butil::fast_rand()) % memClusterMap_.size();
+        auto iter = memClusterMap_.cbegin();
+        for (int i = 0; i <= randId; ++i, ++iter) continue;
```
Comment (cw123) [inline curvefs/src/mds/topology/topology.cpp:1738]: for condition is wrong

File: curvefs/src/client/rpcclient/mds_client.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (SeanHai) [reply curvefs/src/client/rpcclient/mds_client.cpp:435]: don't use response.cluster() instead with *response.mutable_cluster()

File: curvefs/src/mds/topology/topology_manager.cpp (hunk 11 above)
Comment (Cyber-SiKu) [reply curvefs/src/mds/topology/topology_manager.cpp:1364]: > add readlock registMemcacheClusterMutex_

In `topology_->ListMemcacheClusters();` there is `ReadLockGuard rlockMemcacheCluster(memcacheClusterMutex_);` which can already guarantee the validity of the data?

File: curvefs/src/mds/topology/topology_manager.cpp (hunk 12 above)
Comment (Cyber-SiKu) [reply curvefs/src/mds/topology/topology_manager.cpp:1379]: > add readlock registMemcacheClusterMutex_

ditto
`topology_->AllocOrGetMemcacheCluster` to ensure data consistency.

File: curvefs/src/mds/topology/topology.cpp (hunk 13 above)
Comment (Cyber-SiKu) [reply curvefs/src/mds/topology/topology.cpp:1711]: > two line is dup

fix

File: curvefs/src/mds/topology/topology.cpp (hunk 14 above)
Comment (Cyber-SiKu) [reply curvefs/src/mds/topology/topology.cpp:1738]: > for condition is wrong

fix

**Comment by YunhuiChen** [comment]
recheck

File: curvefs/src/client/rpcclient/mds_client.cpp (hunk 3 above)
Comment (Cyber-SiKu) [reply curvefs/src/client/rpcclient/mds_client.cpp:418]: > add metric for this

fix

**Comment by SeanHai** [comment]
create a new issue related this pr about add unit tests of this pr.

**Comment by Cyber-SiKu** [comment]
> create a new issue related this pr about add unit tests of this pr.

add #2147

**Comment by Cyber-SiKu** [comment]
recheck

**Comment by Cyber-SiKu** [comment]
check

**Comment by Cyber-SiKu** [comment]
cicheck

**Review by cw123 — APPROVED** [review]

**Review by SeanHai — APPROVED** [review]

**Comment by Cyber-SiKu** [comment]
cicheck

**Comment by Cyber-SiKu** [comment]
cicheck

**Pull request merged**